fennel/datasets/datasets.py

```diff
@@ -410,44 +453,113 @@ def dsschema(self):
         )
 
 
+class UDFType(str, Enum):
+    expr = "expr"
+    python = "python"
+
+
 class Assign(_Node):
     def __init__(
-        self, node: _Node, column: str, output_type: Type, func: Callable
+        self,
+        node: _Node,
+        column: Optional[str],
+        output_type: Optional[Type],
+        func: Optional[Callable],
+        **kwargs,
     ):
         super().__init__()
         self.node = node
         self.node.out_edges.append(self)
         self.func = func
         self.column = column
         self.output_type = output_type
+        self.output_expressions = {}
+        self.assign_type = UDFType.python
+        # Check that either node, column and output_type are all None or all not None
+        if (
+            node is None
+            and column is None
+            and output_type is None
+            and len(kwargs) == 0
+        ) or (
+            node is not None
+            and column is not None
+            and output_type is not None
+            and len(kwargs) > 0
+        ):
+            raise ValueError(
+                "Assign expects either to use the arguments `node`, `column` and `output_type` or use the keyword arguments for expressions"
+            )
+        # Map of column names to expressions
+        if len(kwargs) > 0:
+            for k, v in kwargs.items():
+                self.output_expressions[k] = v
+            self.assign_type = UDFType.expr
+
+    @classmethod
+    def from_expressions(cls, self, **kwargs):
+        for k, v in kwargs.items():
+            if not isinstance(v, TypedExpr):
+                raise ValueError(
+                    "Assign.from_expressions expects all values to be of type Expr",
+                    f"found `{type(v)}` for column `{k}`",
+                )
```

Comment:
Nit but you might want to add a special check where it's not `TypedExpr` but is `Expr` and give an error message that is more specific.